File: interpreter-for-cpp/src/Runtime/Sentence/SentenceEnumDefine.cpp
#include "SentenceEnumDefine.h"
#include "../Value/ValueTool.h"
#include "../Variable.h"

using namespace peak::interpreter;

SentenceEnumDefine::SentenceEnumDefine(const std::string& name, const std::list<std::pair<std::string, std::shared_ptr<ValueNumber>>>& valueList)
	: _name(name), _valueList(valueList) {
}

ExecuteResult SentenceEnumDefine::Execute(std::shared_ptr<Space> space) {
	auto valueObject = std::shared_ptr<ValueObject>(new ValueObject());
	auto variable = std::shared_ptr<Variable>(new Variable(_name, VariableAttribute::Const, valueObject));
	if (!space->AddVariable(variable)) {
		ErrorLogger::LogRuntimeError(_name);
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::EnumDefine, "The enum name \"" + _name + "\" is exist!");
		return ExecuteResult::Failed;
	}
	auto valueSpace = valueObject->GetSpace();
	for (auto& pair : _valueList) {
		auto valueVariable = std::shared_ptr<Variable>(new Variable(pair.first, VariableAttribute::Const, pair.second));
		if (!valueSpace->AddVariable(valueVariable)) {
			ErrorLogger::LogRuntimeError(pair.first);
			ErrorLogger::LogRuntimeError(_name);
			ErrorLogger::LogRuntimeError(ErrorRuntimeCode::EnumDefine, "The item name \"" + _name + "\" is exist!");
			return ExecuteResult::Failed;
		}
	}
	return ExecuteResult::Successed;
}

File: interpreter-for-cs/src/Runtime/Sentence/SentenceEnumDefine.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

using System.Collections.Generic;

namespace peak.interpreter {

	public class SentenceEnumDefine : Sentence {
		private string _name = "";
		private LinkedList<KeyValuePair<string, ValueNumber>> _valueList = null;
		public SentenceEnumDefine(string name, LinkedList<KeyValuePair<string, ValueNumber>> valueList) {
			_name = name;
			_valueList = valueList;
		}
		public override ExecuteResult Execute(Space space) {
			var valueObject = new ValueObject();
			var variable = new Variable(_name, VariableAttribute.Const, valueObject);
			if (!space.AddVariable(variable)) {
				ErrorLogger.LogRuntimeError(_name);
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.EnumDefine, "The enum name \"" + _name + "\" is exist!");
				return ExecuteResult.Failed;
			}
			var valueSpace = valueObject.space;
			foreach (var pair in _valueList) {
				var valueVariable = new Variable(pair.Key, VariableAttribute.Const, pair.Value);
				if (!valueSpace.AddVariable(valueVariable)) {
					ErrorLogger.LogRuntimeError(pair.Key);
					ErrorLogger.LogRuntimeError(_name);
					ErrorLogger.LogRuntimeError(ErrorRuntimeCode.EnumDefine, "The item name \"" + _name + "\" is exist!");
					return ExecuteResult.Failed;
				}
			}
			return ExecuteResult.Successed;
		}

	};
} // namespace peak.interpreter